### ISSUE TITLE
set and intro fail when overlapping generated names

### ISSUE THREAD

**Issue by jashug** [coq/coq#6577]
#### Version

Coq 8.7.0

#### Description of the problem

In regards to the criterion [here](https://github.com/coq/coq/pull/268#issuecomment-243842175), I don't see how it is tenable for the `set` or `intro` tactics to fail here.
```
Goal True -> True -> Set.
intro.
Fail set (H := False).
Fail intro H.
```

`intro` auto-generates the name `H`, which then causes `set` to fail. This direction is fine,
but if you want to write scripts that are robust to changes in auto-generated names,
you can't use `set` or `intro` with an explicit name when you have hypotheses with auto-generated names in context.

To be able to write robust scripts,
I would expect both `set` and `intro` to succeed here, alpha-renaming the previous `H` away.

A slightly different example:
```
Goal True -> True.
intro; set (H0 := I); exact H0.
Qed.

Axiom H : True.
Goal True -> True.
Fail intro; set (H0 := I); exact H0.
```

Adding `H` to the global context causes `intro` to choose `H0` for the generated name, which causes `set` to fail.

**Comment by ppedrot** [comment]
Note that you can write robust scripts by never ever writing explicit identifiers (as you did here with `H`) and use instead the `fresh` function. I think this is the currently preferred solution.

**Comment by jashug** [comment]
Oh, so my second example would look something like
```coq
Axiom H : nat.
Goal True -> True.
let H := fresh in
intro H; exact H.
Qed.
```
That is clever. Is there a way to bind fresh names over multiple statements?
(splitting with a period `intro H. exact H.`)
If not, this seems a difficult style to recommend.

**Comment by Zimmi48** [comment]
> If not, this seems a difficult style to recommend.

It's clearly not a style to recommend IMHO. Alpha-renaming (or clearing) auto-generated names looks much saner to me.

**Comment by SkySkimmer** [comment]
```coq
Section foo.
  Variable x : nat.

  Goal bool -> x = x.
  Proof.
    refine (fun x => _).
    (* x : nat, x0 : bool |= x = x *)
```

**Comment by jashug** [comment]
@ppedrot I'm curious as to what Ltac2 does with the above. Does refine succeed in claiming the name `x`?

**Comment by Zimmi48** [comment]
All these examples leave me with the feeling that this whole naming thing was never really thought through or specified. It would be good to write down some specifications and then ensure that all tactics / all mechanisms are following these specifications.

**Comment by ppedrot** [comment]
@SkySkimmer This is a dubious behaviour that should be thought of as a work-around to the section mechanism being broken, and this is somewhat unrelated to the actual tactic language. Unluckily, there is no other easy way currently. In particular, @jashug Ltac2 has to do the very same thing.

**Comment by jashug** [comment]
I am recording here a reply by Pierre Courtieu (possibly @Matafou) to a question on Coq Club:
https://sympa.inria.fr/sympa/arc/coq-club/2018-01/msg00139.html

> About tactics updating specifically I came up to the conclusion that one should
> 
> 1) Never name hypothesis by hand in "intros" or "as" clauses. Because
> if you have explicit "intros foo bar" or "inversion h as [ foo bar]"
> then if you change definitions you need to change all this intros and
> as everywhere. Scripts are supposed to be accepted (except where
> changes do impact things) once you have done that, but that is already
> very annoying.
> 
> 2) Never use coq's default naming scheme as it is not stable when
> changing definition.
> 
> So I came up with an automatic naming scheme that, if correctly
> handled is almost stable by definition changes
> (http://cedric.cnam.fr/~courtiep/downloads/LibHypsNaming.v). There is
> some boilerplate anyway: it requires to maintain a naming function in
> Ltac by hand.
> 
> Something like cortouche may be even better
> (https://github.com/pedagand/cortouche).

This brings up the interesting point that robust scripts are also expected to be resilient to small changes in relevant definitions, not just changes in automatically chosen names.

**Comment by Matafou** [comment]
Hi there,
While I am quite happy with my advice naming stuff I am willing to give cortouche a real try but Pierre-Evariste told me that since it seems doable in ltac it is not worth maintaining it as a plug-in. Is there any ltac development of cortouche around?

**Comment by Matafou** [comment]
Because ideally this just makes hypothesis autonaming something of the past.

**Comment by Zimmi48** [comment]
@Matafou Here is a way of doing it in Ltac (courtesy of @jonleivent):

```coq
Class Hyp (T : Type) := hyp : T.
Hint Extern 10 (Hyp _) => exactly_once multimatch goal with H : _ |- _
=> exact H end : typeclass_instances.
Class AHyp (T : Type) := ahyp : T.
Hint Extern 10 (AHyp _) => exactly_once multimatch goal with H : _ |- _
=> eapply H end : typeclass_instances.

Parameter is_bar : nat -> nat -> Prop.

Lemma example_with_evar_names :
   is_bar 1 0 -> is_bar 1 1 -> is_bar 0 1 -> exists n1 n2, is_bar n1 n2
/\ is_bar n1 n2.
Proof.
   intros.
   do 2 eexists; split.
   exact (hyp: is_bar ?[X] ?X). (*second subgoal is now is_bar 1 1*)
   exact (hyp: is_bar 1 1).
Qed.
```

(the example is there to show that it works even with non-linear patterns)

**Comment by Matafou** [comment]
Neat!
It won't work directly with tactics expecting a name (like inversion):

```
Lemma foo: forall x y: nat,  Some x = Some y -> x = y .
Proof.
  intros x y H.
  inversion (hyp: _ = _). (* [tactic:quantified_hypothesis] expected after 'inversion'  *)
```

but that is probably easy to fix, I am not expert enough in ltac, I only found this painful workaround:

```
Ltac name_of t :=
  match t with
  | hyp => 
    let u := type of t in
    match goal with
      h:u |- _ => h
    end
  end.

Lemma foo: forall x y: nat,  Some x = Some y -> x = y .
Proof.
  intros x y H.
  let h := name_of (hyp: _ = _) in inversion h.
```

Maybe I should start another PR to add a clean version of this in tactics.v?

**Comment by Zimmi48** [comment]
What's the first `match` for? I found this works as well:

```coq
Ltac name_of t :=
  let u := type of t in
  match goal with
    h:u |- _ => h
  end.
```

and it looks quite clean to me.

**Comment by Matafou** [comment]
you are right, it is even better as it does not trigger one more typeclass resolution.

But still: writing `let h := name_of (hyp: _ = _) in inversion h.` instead of `inversion (hyp: _ = _)` remains painful.
P.

**Comment by Zimmi48** [comment]
So it's still more painful than what cortouche allows. I don't think that there is anything against having this done in a plugin BTW, other than it's harder to install. But cortouche could possibly be shipped with Coq if it's got users. It's so light and it would be easy to maintain.

**Comment by Matafou** [comment]
I am quite sure @jonleivent can solve this problem with tactics taking names as arguments. Maybe with a tactic notation? Otherwise indeed it might be interesting to ship cortouche with coq. Actually cortouche might be even lighter to implement if it relies on Jonathan's ltac, the only thing that needs to be done in ocaml is to insert `name_of` when needed.

**Comment by Matafou** [comment]
I did not manage to make this ltac support names as arguments. @jonleivent do you think it is possible?
